### ISSUE TITLE
Resource.use isn't stack safe

### ISSUE THREAD

**Issue by andimiller** [typelevel/cats-effect#469]
When implementing `mapK` for `Resource` I went to test it to see if it was stack-safe, then instantly ran into `use` not being stack safe, I expected the following to work:
```scala
test("use should be stack-safe over binds") {
  (1 to 10000).foldLeft(Resource.liftF(IO.unit)) { case (r, _) =>
    r.flatMap(_ => Resource.liftF(IO.unit))
  }.use { r =>
    IO.unit
  }.unsafeRunSync shouldBe ()
}
```

this fails with 
```
[info] cats.effect.ResourceTests *** ABORTED ***
[info]   java.lang.StackOverflowError:
[info]   at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:134)
[info]   at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
[info]   at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
[info]   at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
[info]   at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
[info]   at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
[info]   at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:372)
[info]   at cats.effect.internals.IORunLoop$RestartCallback.apply(IORunLoop.scala:312)
[info]   at cats.effect.internals.IORunLoop$.cats$effect$internals$IORunLoop$$loop(IORunLoop.scala:136)
[info]   at cats.effect.internals.IORunLoop$RestartCallback.signal(IORunLoop.scala:351)
[info]   ...
```

**Comment by andimiller** [comment]
@alexandru cool, that makes sense to me, I will close this since it's not an actual issue then

**Comment by kaishh** [comment]
@alexandru @SystemFw 
What about post-processing of resources or chained allocations? e.g. I have a usecase where I want to allocate a long-living file, try to dump the buffer into it, and if either of these steps fail - delete the file and keep the buffer in memory. This is naturally expressible as flatMaps on `Resource`, but since I have a very large number of files, I have to keep in mind that `Resource` is stack-unsafe.

In this particular case, I'm still using `Resource`, since ZIO's very low auto-fork threshold resets the stack for me, but if I were using another effect type I wouldn't be able to get away with it that easily. Maybe another flatMap constructor could be added that _doesn't_ attach an empty finalizer to avail worries about heap overflow?

**Comment by alexandru** [comment]
@kaishh The stack depth is pretty big, it's impossible to hit it with a sample like yours.

Also note that auto-forking behavior is an implementation detail that's orthogonal to the stack safety of `bracket`.

**Comment by kaishh** [comment]
@alexandru I'm using 3 flatMaps per file, plus .traverse over 5000+ files, so I'm clearly going over 10000 flatMaps quite fast

**Comment by alexandru** [comment]
So you're opening 5000 file handles at the same time? That seems like a waste and why would you need them in the same run-loop, at the same time anyway? What's the use-case?

That said, the `Resource` implementation is probably fine. `flatMap` is just building a data structure, but it gets translated into `bracket` calls on `use` and `bracket` is the actual issue.

This is an implementation detail of Cats-Effect's `IO` and its `bracket`. I remember this being a conscious design decision on my part actually, because it has better performance and stack overflow errors are for me a good feature.

I can be persuaded however and the fix would be easy.

**Comment by kaishh** [comment]
I'm creating temporary files, not handles, handles are closed immediately after writing, the point is to serialize and dump a large amount of data on disk in constant space - serialized data is denormalized in multiple variants, it no longer fits in-memory - upload and delete all temporaries.

Ok, I guess the issue is misnamed then. My personal opinion is that a standard IO type shouldn't be too opinionated, and should be more or less safe for all meaningful uses - and it may sometimes be meaningful to open a huge amount of brackets.

**Comment by t3hnar** [comment]
Hello guys.

I recently faced the same `java.lang.StackOverflowError` just because we are using `Resource` as default building type for composing our services within application. I believe that this is not a wrong usage. And looks like we are approaching stack limit and randomly getting `StackOverflowError`. 

What do you think about reconsidering the status of this issue and mark it is as **bug** and somebody could try to fix this?

Meanwhile we made workaround: wrapping into `Resource(resource.allocated)` on per some logical part/module, this breaks the chain of huge `Resource.flatMap` calls and safes us from getting `StackOverflowError`

**Comment by SystemFw** [comment]
It would be worth testing this with 3.0.0-M1, which has a different `bracket` implementation

**Comment by t3hnar** [comment]
@SystemFw 
I had a unit test covering this case, it started to fail after bumping to `2.2.0` version - means something is fixed or improved even before `3.0.0-M1`

**Comment by mberndt123** [comment]
Hey @alexandru,

I'm using the `Resource` type in conjunction with JPL, the Java Prolog Bridge. I generate Prolog clauses dynamically and add them to the Prolog database using the [`assertz/2`](https://www.swi-prolog.org/pldoc/man?predicate=assert/2) predicate. To avoid leaking memory, I need to call [`erase/1`](https://www.swi-prolog.org/pldoc/man?predicate=erase/1) once I no longer need them. Since everything I add to the database eventually needs to be erased even when an error occurs, a `Resource` seems like a perfect fit. So if you're still looking for a persuasive example for when it's reasonable to have thousands of nested resources, I think this is one. It's not unusual to have hundreds or thousands of clauses in a Prolog database.
